src/plugins/omemo-views/utils.js

```diff
@@ -0,0 +1,154 @@
+        return tpl_video(obj_url);
+    } else {
+        return tpl_file(obj_url, uri.filename());
+    }
+
+}
+
+// TODO: move the non-template part to headless
+function addEncryptedFiles(text, offset, richtext) {
+    const objs = [];
+    try {
+        const parse_options = { 'start': /\b(aesgcm:\/\/)/gi };
+        URI.withinString(
+            text,
+            (url, start, end) => {
+                objs.push({ url, start, end });
+                return url;
+            },
+            parse_options
+        );
+    } catch (error) {
+        log.debug(error);
+        return;
+    }
+    objs.forEach(o => {
+        const uri = getURI(text.slice(o.start, o.end));
+        const promise = getAndDecryptFile(uri)
+            .then(obj_url => getTemplateForObjectURL(uri, obj_url, richtext));
+
+        const template = html`${until(promise, '')}`;
+        richtext.addTemplateResult(o.start + offset, o.end + offset, template);
+    });
+}
+
+// TODO: move non-UI functions (eg. el.model stuff) to headless
+export function onChatInitialized (el) {
+    el.listenTo(el.model.messages, 'add', message => {
+        if (message.get('is_encrypted') && !message.get('is_error')) {
+            el.model.save('omemo_supported', true);
+        }
+    });
+    el.listenTo(el.model, 'change:omemo_supported', () => {
+        if (!el.model.get('omemo_supported') && el.model.get('omemo_active')) {
+            el.model.set('omemo_active', false);
+        } else {
+            // Manually trigger an update, setting omemo_active to
+            // false above will automatically trigger one.
+            el.querySelector('converse-chat-toolbar')?.requestUpdate();
+        }
+    });
+    el.listenTo(el.model, 'change:omemo_active', () => {
+        el.querySelector('converse-chat-toolbar').requestUpdate();
+    });
+}
+
+// TODO: move non-UI part of this into headless
+function toggleOMEMO (ev) {
+    ev.stopPropagation();
+    ev.preventDefault();
+    const toolbar_el = u.ancestor(ev.target, 'converse-chat-toolbar');
+    if (!toolbar_el.model.get('omemo_supported')) {
+        let messages;
+        if (toolbar_el.model.get('type') === _converse.CHATROOMS_TYPE) {
+            messages = [
+                __(
+                    'Cannot use end-to-end encryption in this groupchat, ' +
+                        'either the groupchat has some anonymity or not all participants support OMEMO.'
+                )
+            ];
+        } else {
+            messages = [
+                __(
+                    "Cannot use end-to-end encryption because %1$s uses a client that doesn't support OMEMO.",
+                    toolbar_el.model.contact.getDisplayName()
+                )
+            ];
+        }
+        return api.alert('error', __('Error'), messages);
+    }
+    toolbar_el.model.save({ 'omemo_active': !toolbar_el.model.get('omemo_active') });
+}
+
+// TODO: move non-UI part of this into headless (but is there any?)
```

Comment:
This one I think is purely UI so I'm going to leave it here untouched :slightly_smiling_face: